text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+3. The last option would be to use a manually constructed block (more details below).
+
+A desirable goal here would be to revert as few transactions from all the forks as possible, in order to limit the damage. Every transaction that was included in a block which appeared to be finalized for a time could have caused real world consequences (like, for example, a merchant sending ordered goods). Reverting any such transaction might potentially cause a loss to a user.
+
+Option 1 essentially reverts all transactions that happened after the block whose FTT has been exceeded. Option 2 attempts to revert the transactions from all forks but one. However, it is possible that some of the transactions on different forks of the chain don't conflict with each other. In such a case, we could attempt to replay all of them on the new trusted branch. This is the idea behind option 3: to gather the non-conflicting transactions, create a block containing them and collectively declare it correct, outside of the consensus algorithm.
+
+After the users collectively decide upon the last correct block, the transactions to be replayed and the list of equivocators to be slashed, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network could be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators, which is based on the bonded validators of the last correct block, with the equivocators removed.
+- There is a different initial global state: it is not empty anymore, but it is the post-state of the last correct block.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The information necessary for restarting the network can (and probably should) be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was.
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the last correct block, the transactions that need to be replayed on top of it and the set of equivocating validators (exchanging evidence if necessary).
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
```

Comment:
What is the risk here around 'tweaking' or 'messing with' this chainspec?  I don't know if there are any good solutions, seems like this needs to be tightly accounted for via governance rules?

Reply:
I think once the equivocation catastrophe has happened, in a sense all bets are off.
Because if we defined a strict governance process, with strict rules and a strict quorum, then too many actors of that off-chain process could be malicious, too. Anything that can be precisely defined can be implemented, and is subject to equivocation catastrophes itself.

Social consensus allows the system to survive even such a situation, just like it can survive an equivocation catastrophe: Because _anyone_ can continue the chain, with _any_ set of validators, and what matters is only whether the public accepts that as the "real" new chain.

Reply:
A good question. I would expect that someone messing with their chainspec locally would mostly hurt themselves, because their node software would expect a different initial state of the network than everyone else's. It could be a problem if a majority of the validators collude in order to make some changes to the chainspec which others didn't agree to... But that would probably mean them effectively creating a hard fork (or are there possibilities of messing with the chainspec that would go unnoticed by other nodes for some time and create problems with some delay? cc @afck  @goral09 )

Reply:
I agree: Changing the chain spec means creating a hard fork.
If the others are not following your fork, it's useless and not dangerous.
If the others _are_ following your fork, you successfully reached social consensus, and it's the accepted fork now.